avm/src/vm/memory.rs

```diff
@@ -1,654 +1,390 @@
-      panic!("Tried to create an array at address {}, but one already exists.", addr);
-    }*/
-    //println!("create fractal: @{}", addr);
-    self.write_fractal_mem(addr, &[]);
+  pub fn push_fixed(self: &mut HandlerMemory, addr: i64, val: i64) {
+    let mem = self.read_mut_fractal(addr);
+    mem.push((std::usize::MAX, val));
   }
 
-  pub fn push_fractal_fixed(self: &mut HandlerMemory, addr: i64, val: i64) {
-    // This implementation uses the `mem` vector as a way to keep track of the total length of the
-    // array, as well. It's simple but wastes space when the inserted value is variable-length
-    // (such as other strings or other arrays), however it greatly simplifies addressing and
-    // lookup, particularly for `Array<any>`, which is also what user-defined types are transformed
-    // into. In the future we could have an address translation layer and pack the data as tightly
-    // as we can, assuming that doesn't impose a large performance penalty, while this simple
-    // solution only adds an extra key's worth of space usage, but does have memory copy issues due
-    // to the constant resizing.
-    let arr = self.get_mut_fractal(addr);
-    let idx = arr.mem.len();
-    arr.mem.push(0);
-    arr.either_mem.push(-1);
-    arr.registers_ish.push(0);
-    arr.write_fixed(idx as i64, val);
-    //println!("push fixed: @{}[{}]: {}", addr, idx, val);
+  pub fn push_fractal(self: &mut HandlerMemory, addr: i64, val: &Vec<(usize, i64)>) {
+    let a = self.mems.len();
+    let mem = self.read_mut_fractal(addr);
+    mem.push((a, std::usize::MAX as i64));
+    self.mems.push(val.clone());
   }
 
-  pub fn push_nested_fractal_mem(self: &mut HandlerMemory, addr: i64, val: Vec<i64>) {
-    let arr = self.get_mut_fractal(addr);
-    let idx = arr.mem.len() as i64;
-    arr.mem.push(0);
-    arr.either_mem.push(idx);
-    arr.registers_ish.push(0);
-    arr.write_fractal_mem(idx, &val);
-    //println!("push nested mem: @{}[{}]: {}", addr, idx, val[0]);
+  pub fn push_register(self: &mut HandlerMemory, addr: i64, other_addr: i64) {
+    let (a, b) = self.addr_to_idxs(other_addr);
+    // Special path for strings in global memory
+    if a == 0 {
+      let strmem = self.mems[0][b..].to_vec().clone();
+      let new_a = self.mems.len();
+      self.mems.push(strmem);
+      let mem = self.read_mut_fractal(addr);
+      mem.push((new_a, std::usize::MAX as i64));
+    } else {
+      let mem = self.read_mut_fractal(addr);
+      mem.push((a, b as i64));
+    }
   }
 
-  pub fn push_nested_fractal(self: &mut HandlerMemory, addr: i64, val: HandlerMemory) {
-    let arr = self.get_mut_fractal(addr);
-    let idx = arr.fractal_mem.len() as i64;
-    //println!("push nested: @{}[{}]: {}", addr, arr.mem.len(), val);
-    arr.mem.push(0);
-    arr.either_mem.push(idx);
-    arr.fractal_mem.push(val);
-    arr.registers_ish.push(0);
+  pub fn push_idxs(self: &mut HandlerMemory, addr: i64, a: usize, b: usize) {
```

Comment:
nit push_addr_idxs ?

Reply:
No, all of the `push_<thing>` functions are pushing into some nested memory. This function is for pushing the virtual pointer itself, which made certain operations faster *and* simpler. Unless you think they should *all be `push_addr_<thing>`?

Reply:
since it is a public function this particular method does seem confusing since idx could mean anything so a part of my does want to go with `push_addr_<thing>` for all, but more of a nit. fine leaving it as-is with a comment